### Dinesh Rajan 
##### 20CSEG06

# LINEAR PROGRAMMING PROBLEM(SIMPLEX)

## A catering company is to make lunch for a business meeting. It will serve ham sandwiches, light ham sandwiches, and vegetarian sandwiches. A ham sandwich has 1 serving of vegetables, 4 slices of ham, 1 slice of cheese, and 2 slices of bread. A light ham sandwich has 2 serving of vegetables, 2 slices of ham, 1 slice of cheese and 2 slices of bread. A vegetarian sandwich has 3 servings of vegetables, 2 slices of cheese, and 2 slices of bread. A total of 10 bags of ham are available, each of which has 40 slices; 18 loaves of bread are available, each with 14 slices; 200 servings of vegetables are available, and 15 bags of cheese, each with 60 slices, are available. Given the resources, how many of each sandwich can be produced if the goal is to maximize the number of sandwiches?

### Solution:
We wish to maximize the number of sandwiches, so let:

x = # of ham sandwiches

y = # of light ham sandwiches

z = # of vegetarian sandwiches

The total number of sandwiches is given by

S = x + y + z

The constraints will be given by considering the total amount of ingredients available. That is, the company has a limited amount of ham, vegetables, cheese, and bread.

In total, the company has
10(40) = 400 slices of ham, 18(14) = 252 slices of bread, 200 servings of vegetables, and 15(60) = 900 slices of cheese available. At most, the company can use the above amounts.

There are two sandwiches that use ham—the first requires 4 slices of ham and the second requires only 2, per sandwich. That is, 4x + 2y ≤ 400

That is, the total number of slices of ham cannot exceed 400.

Each sandwich requires 2 slices of bread so 2x + 2y + 2z ≤ 252

The ham sandwiches have 1 and 2 servings of vegetables, respectively, and the vegetarian sandwich has 3 servings of vegetables. So, 1x + 2y + 3z ≤ 200

Both ham sandwiches require one slice of cheese, and the vegetarian sandwich requires two slices of cheese, so, 1x + 1y + 2z ≤ 900   Below is the completed linear programming model for this example.

Maximize: S = x + y + z
Subject To: 4x + 2y ≤ 400 , 
            2x + 2y + 2z ≤ 252 , 
             x + 2y + 3z ≤ 200 , 
            1x + 1y + 2z ≤ 900 , 
             x, y, z ≥ 0

These constraints satisfy the requirements for the simplex method, so we proceed.

Incorporating slack variables, we get:

4x + 2y + 0z + s1 = 400

2x + 2y + 2z + s2 = 252

x + 2y + 3z + s3 = 200

x + y + 2z + s4 = 900

–x – y – z + S = 0

### THE PROBLEM IS SOLVED USING PYTHON 

### Using SCIPY

In [1]:
from scipy.optimize import linprog

In [2]:
obj=[1,1,1]
lhs_ineq=[[4,2,0],[2,2,2],[1,2,3],[1,1,2]]
rhs_ineq=[400,252,200,900]
bnd=[(0,float("inf")),(0,float("inf")),(0,float("inf"))]

In [3]:
opt = linprog(c=obj, A_ub=lhs_ineq,b_ub=rhs_ineq,bounds=bnd,method="simplex")

In [4]:
opt

     con: array([], dtype=float64)
     fun: 0.0
 message: 'Optimization terminated successfully.'
     nit: 6
   slack: array([400., 252., 200., 900.])
  status: 0
 success: True
       x: array([0., 0., 0.])

In [5]:
opt.fun

0.0

In [6]:
opt = linprog(c=obj, A_ub=lhs_ineq,b_ub=rhs_ineq,bounds=bnd,method="revised simplex")

In [7]:
opt

     con: array([], dtype=float64)
     fun: 0.0
 message: 'Optimization terminated successfully.'
     nit: 0
   slack: array([400., 252., 200., 900.])
  status: 0
 success: True
       x: array([0., 0., 0.])

## Using PuLP package

In [8]:
from pulp import LpMaximize, LpProblem, LpStatus, lpSum, LpVariable

In [9]:
# define the model
model = LpProblem(name="resource-allocation", sense=LpMaximize)

#define the variable
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}

# Add constraints
model += (lpSum(x.values()) <= 400,"ham")
model += (2 * x[1] + 2 * x[2] + 2 * x[3] <= 252,"vegetables")
model += (x[1] + 2 * x[2] + 3 * x[3] <= 200,"cheese")
model += (x[1] + x[2]+ 2 * x[3] <= 900,"bread")

# Set the objective
model +=  x[1] + x[2] + x[3]

# Solve the optimization problem
status = model.solve()

# Get the results
print(f"status: {model.status}, {LpStatus[model.status]}")
print(f"objective: {model.objective.value()}")

for var in x.values():
    print(f"{var.name}: {var.value()}")

for name, constraint in model.constraints.items():
    print(f"{name}: {constraint.value()}")

status: 1, Optimal
objective: 126.0
x1: 52.0
x2: 74.0
x3: 0.0
ham: -274.0
vegetables: 0.0
cheese: 0.0
bread: -774.0


We now have the optimal solution

x=100 (basic variable in row 1) , 
y=0 (nonbasic variable) , 
z=26 (basic variable row 2) , 
s1=0 (nonbasic variable) , 
s2=0 (nonbasic variable) , 
s3=22 (basic variable row 3) , 
s4=748 (basic variable row 4) , 
S=126 (basic variable row 5) , 
Of course we are really just interested in: x=100, y=0, z=26, S=126

### We find that 100 ham sandwiches, 26 vegetarian sandwiches, and 0 light ham sandwiches should be made to maximize the total number of sandwiches made.